File: velometro/velometro/page/bom_costing/bom_costing.py
# Copyright (c) 2017, ERPNext and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from erpnext.stock.doctype.item.item import get_last_purchase_details
from erpnext.accounts.doctype.pricing_rule.pricing_rule import get_pricing_rule_for_item

from frappe.utils import flt

def sum_bom_cost(item_list):
	total_cost = 0
	for item in item_list:
		total_cost += item.unit_cost * item.qty

	return total_cost
	
def get_item_cost(item_code, cost_method, qty):
	item_price = 0
	
	if cost_method == "Item Valuation Rate":
		# Get the item valuation method
		item_price == frappe.db.get_value("Bin",{"item_code": "item_code"}, "valuation_rate")

	elif cost_method == "Last Purchase Rate":
		last_purchase_details = get_last_purchase_details(item_code)
		item_price = last_purchase_details.base_rate
		
	elif cost_method == "Default Supplier":
	
		item = frappe.get_doc("Item", item_code)
		args = {
			"doctype": item.doctype,
			"name": item.name,
			"item_code": item.item_code, 
			"transaction_type": "buying",
			"supplier": item.default_supplier,
			"qty": qty
		}
		args = frappe._dict(args) 
		
		#frappe.msgprint(str(args))
		pr_price = 0		
		pr_result = get_pricing_rule_for_item(args) 
		if pr_result.pricing_rule:
			pricing_rule = frappe.get_doc("Pricing Rule", pr_result.pricing_rule)
			pr_price = pricing_rule.price 
		else:
			#Need to find the item price
			price_list = frappe.db.get_value("Supplier", item.default_supplier, "default_price_list")
			item_price = frappe.db.get_value("Item Price", {"price_list": price_list,"item_code": item.item_code}, "price_list_rate") or 0
	
		
	return item_price
	
@frappe.whitelist()
def solve_bom_cost(bom_name, primary_cost, secondary_cost, qty):
	item_list = []
	frappe.msgprint(str(bom_name))
	if bom_name:
		# Get the BOM 
		bom = frappe.get_doc("BOM", bom_name)
		if bom:
			# Get the pricing for each row
			for item in bom.items:
				# Find out the material request type
				mr_type = frappe.db.get_value("Item", item.item_code,"default_material_request_type")
				my_item = {}
				if mr_type == "Manufacture":
					# Check to see if there is a BOM and get the price of that BOM
					if item['bom_no']:
						my_item['unit_cost'] = sum_bom_cost(solve_bom_cost(item.bom_no, primary_cost, secondary_cost, item.qty * qty))
				else:
					# Go through the primary and secondary cost options
					my_item['unit_cost'] = get_item_cost(item.item_code, primary_cost, flt(item.qty) * flt(qty))
					if my_item['unit_cost'] == 0:
						my_item['unit_cost'] = get_item_cost(item.item_code, secondary_cost, flt(item.qty) * flt(qty))
						
				my_item['item_code'] = item.item_code
				my_item['bom_no'] = item.bom_no
				my_item['qty'] = item.qty
				my_item['uom'] = item.stock_uom
				my_item['item_name'] = item.item_name
				item_list.append(my_item)
				
			for op in bom.operations:
				my_item = {}
				my_item['unit_cost'] = op.hour_rate
				my_item['bom_no'] = ""
				my_item['item_code'] = op.operation
				my_item['qty'] = op.time_in_mins / 60
				my_item['uom'] = "Hours"
				my_item['item_name'] = op.description
				item_list.append(my_item)
	
	return item_list
	

File: velometro/velometro/page/bom_costing/bom_costing.js
frappe.pages['bom-costing'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'BOM Costing',
		single_column: true
	});
	
	new erpnext.BOMCosting(wrapper);

	frappe.breadcrumbs.add("Manufacturing");
}

erpnext.BOMCosting = frappe.views.GridReport.extend({
	init: function(wrapper) {
		this._super({
			title: __("BOM Costing"),
			page: wrapper,
			parent: $(wrapper).find('.layout-main'),
			page: wrapper.page
		});

	},
	setup_columns: function() {
		console.log("setup_columns");
		var std_columns = [
			{id: "check", name: "Plot", field: "check", width: 60, hidden: 0,
				formatter: this.check_formatter},
			{id: "item_code", name: "Item Code", field: "item_code", width: 100, hidden: 0,
				formatter: this.text_formatter},
			{id: "item_name", name: "Name", field: "name", width: 300, hidden: 0,
				formatter: this.text_formatter},
			{id: "bom_no", name: "BOM No", field: "bom_no", width: 50, hidden: 1,
				formatter: this.text_formatter},
			{id: "qty", name: "Quantity", field: "qty", plot: false, hidden: 0,
				formatter: this.currency_formatter},
			{id: "unit", name: "UOM", field: "unit", plot: false, hidden: 0,
				formatter: this.text_formatter},
			{id: "item_price", name: "Item Price", field: "item_price", plot: false, hidden: 0,
				formatter: this.currency_formatter},
			{id: "sum_price", name: "Total Price", field: "sum_price", plot: true, hidden: 0,
				formatter: this.currency_formatter}
		];

		this.columns = std_columns;
	},
	filters: [
			
		{fieldtype:"Link", fieldname: "bom", label: __("BOM"),  link:"BOM"},
		{fieldtype:"Select", fieldname: "primary_pricing", label:  __("Primary Pricing Based On"),	
			options:[{label: __("Item Valuation Rate"), value: "Item Valuation Rate"}, {label: __("Last Purchase Rate"), value: "Last Purchase Rate"},{label: __("Default Supplier"), value: "Default Supplier"}]},		
		{fieldtype:"Select", fieldname: "secondary_pricing", label:  __("Backup Pricing Based On"),	
			options:[{label: __("Item Valuation Rate"), value: "Item Valuation Rate"},{label: __("Last Purchase Rate"), value: "Last Purchase Rate"},{label: __("Default Supplier"), value: "Default Supplier"}]},		
		{fieldtype:"Data", fieldname: "quantity", label: __("Build Quantity"), "default": 1}
	],
	setup_filters: function() {
		console.log("setup_filters");
		var me = this;
		this._super();

		this.trigger_refresh_on_change(["bom", "quantity"]);
		
		
	},
	init_filter_values: function() {
		console.log("init_filter_values");
		
		var me = this;
		
		
		this._super();
		this.filter_inputs.primary_pricing.val('Item Valuation Rate');
		this.filter_inputs.secondary_pricing.val('Last Purchase Rate');
		this.filter_inputs.quantity.val('1');
		
		
		
		
	},	
	get_data: function(callback) {
		console.log("get_data");
		var me = this;

		if(!me.setup_filters_done) {
				me.setup_filters();
				me.setup_filters_done = true;
			}
				
		if (this.filter_inputs.bom.val() != ""){
		
			var bname = this.filter_inputs.bom.val();
			var prim = this.filter_inputs.primary_pricing.val();
			var secd = this.filter_inputs.secondary_pricing.val();
			var quant = this.filter_inputs.quantity.val();
			var opts = {
					bom_name: bname,
					primary_cost: prim,
					secondary_cost: secd, 
					qty: quant
				};
			var me = this;
			frappe.call({
				method: "velometro.velometro..page.bom_costing.bom_costing.solve_bom_cost",
				args: opts,
				freeze: true,
				callback: function(r) {
					//Load the data into columns
					console.log(r);
					me.raw_data = r.message;
					callback();
				}
				});
		}
	},
	prepare_data: function() {
		var me = this;

		console.log("prepare_data");
		this.data = [];
		var me = this;
		$.each(this.raw_data, function(i, d) { 
			var new_row = [];
			new_row.item_code = d.item_code;
			new_row.qty = d.qty;
			new_row.item_price = d.unit_cost;
			new_row.name = d.item_name;
			new_row.bom_no = d.bom_no;
			new_row.unit = d.uom;
			new_row.checked = true;
			new_row.id = i;
			var total =  d.qty * d.unit_cost;
			new_row.sum_price = total;
			me.data.push(new_row);
		});
		
	},
	prepare_data_view: function () {
		
		console.log("prepare_data_view");
		//initialize the model
		this.dataView = new Slick.Data.DataView({ inlineFilters: true });
		this.dataView.beginUpdate();
		this.dataView.setItems(this.data);
		if(this.dataview_filter) this.dataView.setFilter(this.dataview_filter);
		if(this.tree_grid.show) this.dataView.setFilter(this.tree_dataview_filter);
		this.dataView.endUpdate();
	},
	setup_chart: function() {
		console.log("setup_chart");
		var me = this;
		var chart_data = this.get_chart_data ? this.get_chart_data() : null;

		console.log(chart_data);
		this.chart = new frappe.ui.Chart({
			wrapper: this.chart_area,
			data: chart_data,
			chart_type: 'pie',
		});
		
	},
	get_chart_data: function() {
		var me = this;
		console.log("get_chart_data");

		var plottable_cols = [];
		console.log(me.data);
		
		var me = this;
		var data = {
			'columns':[],
			onclick: function (d, i) { 
				console.log(me.data[d.index]);
				if (me.data[d.index].bom_no)
				{
					me.filter_inputs.bom.val(me.data[d.index].bom_no);
					me.get_data_and_refresh();
				}
			}
		};
		
		$.each(this.data, function(i, item) {
			console.log('item');
			console.log(item);
			if (item.checked) {
				var item_points = [item.item_code];
				item_points.push(item.sum_price);
				data['columns'].push(item_points);
			}
		});
		return data
	},
	make_filters: function() {
		var me = this;
		$.each(this.filters, function(i, v) {
			v.fieldname = v.fieldname || v.label.replace(/ /g, '_').toLowerCase();
			var input = null;
			if(v.fieldtype=='Select') {
				input = me.page.add_select(v.label, v.options || [v.default_value]);
			} else if(v.fieldtype=="Link") {
				input = me.page.add_data(v.label);
				input.autocomplete({
					source: v.list || [],
				});
			} else if(v.fieldtype==='Button' && v.label===__("Refresh")) {
				input = me.page.set_primary_action(v.label, null, v.icon);
			} else if(v.fieldtype==='Button') {
				input = me.page.add_menu_item(v.label, null, true);
			} else if(v.fieldtype==='Date') {
				input = me.page.add_date(v.label);
			} else if(v.fieldtype==='Label') {
				input = me.page.add_label(v.label);
			} else if(v.fieldtype==='Data') {
				input = me.page.add_data(v.label);
			} else if(v.fieldtype==='Check') {
				input = me.page.add_check(v.label);
			}

			if(input) {
				input && (input.get(0).opts = v);
				if(v.cssClass) {
					input.addClass(v.cssClass);
				}
				/*input.keypress(function(e) {
					if(e.which==13) {
						me.refresh();
					}
				})*/
			}
			me.filter_inputs[v.fieldname] = input;
		});
	}
	
});